conseil-common/src/main/scala/tech/cryptonomic/conseil/common/rpc/RpcClient.scala

```diff
@@ -0,0 +1,176 @@
+package tech.cryptonomic.conseil.common.rpc
+
+import scala.util.control.NoStackTrace
+
+import cats.effect.{Concurrent, Resource}
+import com.typesafe.scalalogging.LazyLogging
+import fs2.Stream
+import org.http4s.{EntityDecoder, EntityEncoder, Header, Method, Uri}
+import org.http4s.client.Client
+import org.http4s.client.dsl.Http4sClientDsl
+
+import tech.cryptonomic.conseil.common.rpc.RpcClient._
+
+/**
+  * JSON-RPC client according to the specification at https://www.jsonrpc.org/specification
+  *
+  * Usage example:
+  *
+  * {{{
+  *   import scala.concurrent.ExecutionContext
+  *   import io.circe.generic.auto._
+  *   import org.http4s.circe.CirceEntityDecoder._
+  *   import org.http4s.circe.CirceEntityEncoder._
+  *
+  *   implicit val contextShift = IO.contextShift(ExecutionContext.global)
+  *
+  *   // Define case class for JSON-RPC method and response
+  *   case class ApiMethodParams(id: Int)
+  *   case class ApiMethodResponse(name: String)
+  *
+  *   // Create stream with the request
+  *   val request = Stream(
+  *     RpcRequest(
+  *       jsonrpc = "2.0",
+  *       mathod = "apiMethod",
+  *       params = ApiMethodParams(id = 1)
+  *       id = "id_of_the_request"
+  *     )
+  *   )
+  *
+  *   // Create JSON-RPC client
+  *   val client = new RpcClient[IO]("endpoint", 10, httpClient)
+  *
+  *   // Call api method with
+  *   val response: Stream[IO, ApiMethodResponse] =
+  *     client.stream[ApiMethodParams, ApiMethodResponse](batchSize = 10)(request)
+  * }}}
+  *
+  * @param endpoint JSON-RPC server url
+  * @param maxConcurrent Max concurrent processes used to fetch data from JSON-RPC server
+  * @param httpClient Http4s http client
+  * @param headers Additional http headers, e.g. authorization
+  */
+class RpcClient[F[_]: Concurrent](
+    endpoint: String,
+    maxConcurrent: Int,
+    httpClient: Client[F],
+    headers: Header*
+) extends Http4sClientDsl[F]
+    with LazyLogging {
+
+  /**
+    * Method that converts [[fs2.Stream]] of JSON-RPC requests into [[fs2.Stream]]] of desired result objects,
+    * or throws the exception with the error. Request [[fs2.Stream]]] can only contain calls to one API method
```

Comment:
Is this method actually going to throw exception? I thought the whole point is to avoid that and return `Either` for example in such cases.